### PR TITLE
"Fixes linear DRIV shape inconsistency"

### PR DESCRIPTION
**Pull request by ssemov**
Addresses the shape inconsistency in the linear DRIV implementation.  

**Changes:**

- I modified the [effect_interval](https://github.com/py-why/EconML/blob/960c0dd832943b8098d50f7b658cea49c822c31c/econml/inference/_inference.py#L284C1-L285C47) computation in LinearModelFinalInference class. This is where the inference method for LinearDRIV comes from.
- I added the modification function - reshape_outcomewise_effects - to utilities. Given an effects matrix, it reshapes the second dimension to be consistent with d_y[0]. 

**Related [issue]**(https://github.com/py-why/EconML/issues/687).  

**Testing:**

- TestDRIV --> OK
- test_inference --> OK
- test_statsmodels --> OK

### CODE REVIEW

**Comment by kbattocchi** [comment]
Thanks for the contribution.  Unfortunately, these changes are causing test failures in `econml.tests.test_dml.TestDML.test_ignores_final_intercept`, `econml.tests.test_discrete_outcome.TestDiscreteOutcome.test_accuracy`, and `econml.tests.test_discrete_outcome.TestDiscreteOutcome.test_accuracy_iv`.

Given that the shapes are as expected for other estimators (like `LinearDML`) using the same inference class, I think the best place to address the problem is probably somewhere in [_dr.py](https://github.com/py-why/EconML/blob/main/econml/iv/dr/_dr.py), although maybe there is a way to address it in the inference class that doesn't break other classes that use it.

**Comment by ssemov** [comment]
Thanks, @kbattocchi! yes, I think the fix I have touches too many things. I'll look into something more local in [_dr.py](https://github.com/py-why/EconML/blob/main/econml/iv/dr/_dr.py).

**Comment by ssemov** [comment]
@kbattocchi I made the change more local and fixed an edge case with the discrete outcome. The new version passes the tests I was failing before.

File: econml/inference/_inference.py [hunk 1]
```diff
@@ -481,6 +482,32 @@ def prefit(self, estimator, *args, **kwargs):
                                                       "their features, but here fit_intercept is True")
         self.model_final.cov_type = self.cov_type
 
+    def effect_interval(self, X, *, T0, T1, alpha=0.05):
+        return self.effect_inference(X, T0=T0, T1=T1).conf_int(alpha=alpha)
```
Comment (kbattocchi) [inline econml/inference/_inference.py:486]: Isn't this identical to what the base class is already doing?

**Comment by ssemov** [comment]
@kbattocchi Removed the redundant effect_interval method.

File: econml/inference/_inference.py [hunk 2]
```diff
@@ -481,6 +482,32 @@ def prefit(self, estimator, *args, **kwargs):
...
+        # We can write effect inference as a function of prediction and prediction standard error of
+        # the final method for linear models
+        X, T0, T1 = self._est._expand_treatments(X, T0, T1)
+        if X is None:
+            X = np.ones((T0.shape[0], 1))
+        elif self.featurizer is not None:
+            X = self.featurizer.transform(X)
+        XT = cross_product(X, T1 - T0)
+        e_pred = reshape_outcomewise_effects(self._predict(XT), self._d_y)
+        e_stderr = reshape_outcomewise_effects(self._prediction_stderr(XT), self._d_y)
```
[the remark this answers is not in the document]
Comment (kbattocchi) [reply econml/inference/_inference.py:495]: (And I think you'll find that without fixing the change in LinearModelFinalInference, other DRIV subclasses like SparseLinearDRIV still suffer from the same issue - please consider modifying [line 457 of test_treatment_featurization](https://github.com/py-why/EconML/blob/960c0dd832943b8098d50f7b658cea49c822c31c/econml/tests/test_treatment_featurization.py#L456-L457) to enable further tests on DRIV subclasses)

**Review by kbattocchi — CHANGES_REQUESTED** [review]
Thanks for trying again, but I don't think StatsModelsInference is the right place for the change - see my comment.

File: econml/inference/_inference.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssemov) [reply econml/inference/_inference.py:495]: @kbattocchi Thanks, Keith, for the help. It makes sense what you are saying, I will work on a fix.

**Comment by ssemov** [comment]
Changes:

- Moves the fix to LinearModelFinalInference
- updates TestTreatmentFeaturization not to skip LinearDRIV and SparseLinearDRIV

Tests:

- TestDML.test_ignores_final_intercept
- TestDiscreteOutcome.test_accuracy 
- TestDiscreteOutcome.test_accuracy_iv 
- TestTreatmentFeaturization

**Review by kbattocchi — APPROVED** [review]
Looks good, thanks!
